banditpylib/learners/collaborative_learner/collaborative_learning.py

```diff
@@ -0,0 +1,390 @@
+      num_pulls += len(arm_feedback.rewards)
+    if self.__central_algo_action_taken:
+      self.__central_algo.update(feedback)
+    elif num_pulls>0:
+      # non-zero pulls not by central_algo => learning step was done
+      for arm_feedback in feedback.arm_feedbacks:
+        if arm_feedback.arm.id == self.__learning_arm:
+          self.__learning_mean = np.array(arm_feedback.rewards).mean()
+          self.__pulls_used = len(arm_feedback.rewards)
+    # else ignore feedback (which is empty)
+
+    self.__central_algo_action_taken = False
+
+  @property
+  def best_arm(self) -> int:
+    # returns arm that the agent chose (could be None)
+    if self.__stage != "termination":
+      raise Exception('%s: I don\'t have an answer yet!' % self.name)
+    return self.__learning_arm
+
+  def broadcast(self) -> Dict[int, Tuple[float, int]]:
+    if self.__stage != "communication":
+      raise Exception('%s: I can\'t broadcast in stage %s!'\
+        % (self.name, self.__stage))
+    return_dict = {}
+    if self.__learning_arm:
+      return_dict[self.__learning_arm] = (self.__learning_mean,
+        self.__pulls_used)
+    self.__complete_round()
+    return return_dict
+
+class LilUCBHeuristicCollaborativeBAIMaster(CollaborativeBAIMaster):
+  r"""Implementation of master in Collaborative Learning Algorithm
+
+  :param int arm_num: number of arms of the bandit
+  :param int rounds: number of rounds of communication allowed
+    (this agent uses one more)
+  :param int horizon: maximum number of pulls the agent can make
+    (over all rounds combined)
+  :param int num_agents: number of agents
+  :param Optional[str] name: alias name
+  """
+
+  def __init__(self, arm_num:int, rounds: int,
+    horizon: int, num_agents: int, name: Optional[str] = None):
+    super().__init__(name)
+    if arm_num <= 1:
+      raise ValueError('Number of arms is expected at least 2. Got %d.' %
+                       arm_num)
+    if rounds <= 1:
+      raise ValueError('Number of rounds is expected at least 2. Got %d.' %
+                       rounds)
+    self.__arm_num = arm_num
+    self.__comm_rounds = rounds - 1
+    self.__T = horizon
+    self.__num_agents = num_agents
+
+  def _name(self) -> str:
+    return "lilUCBHeur_collaborative_master"
+
+  def reset(self):
+    self.__active_arms = list(range(self.__arm_num))
+
+  def __assign_arms(self, agent_ids: List[int]) ->\
+    Dict[int, List[int]]:
+    agent_arm_assignment: Dict[int, List[int]] = {}
+
+    def random_round(x: float) -> int:
+      # if x = 19.3
+      # rounded to 19 with probability 0.7
+      # rounded to 20 with probability 0.3
+      frac_x = x - int(x)
+      if np.random.uniform() > frac_x:
+        return int(x)
+      return int(x) + 1
+
+    if len(self.__active_arms) <= 1:
+      # don't assign any arms if 0/1 active arms
+      agent_arm_assignment = {}
+      for agent_id in agent_ids:
+        agent_arm_assignment[agent_id] = []
+      return agent_arm_assignment
```

Comment:
When there is only one arm, we should still assign arms such that each agent has the same arm.

Reply:
Then how will the agents know that the algorithm has terminated and they should go to "termination" stage too

Reply:
They should monitor the number of rounds used and terminate when it reaches the last round.

Reply:
That is already being done, empty arm assignment is used for early stopping. So if we have only one active arm, then we can terminate the algorithm without doing all the learning steps

Reply:
We could also pass extra variables to `Protocol` in the `elimination` step, and these will be used in `set_input_arms` to decide early stopping

Reply:
Okay, you can remain the current implementation. Also please update the colab whenever the algorithm is changed.

Reply:
Since we are asking the agent for best_arm, this won't work.
I will try using some other method and push the changes